apps/api/lib/api/gateway/socket.ex

```diff
@@ -1,15 +1,54 @@
 defmodule API.Gateway.Socket do
   use Phoenix.Socket
+  alias Domain.Gateways
 
   ## Channels
 
   channel "gateway:*", API.Gateway.Channel
 
   ## Authentication
 
+  def encode_token!(%Gateways.Token{value: value} = token) when not is_nil(value) do
+    body = {token.id, token.value}
+    config = fetch_config!()
+    key_base = Keyword.fetch!(config, :key_base)
+    salt = Keyword.fetch!(config, :salt)
+    Plug.Crypto.sign(key_base, salt, body)
+  end
+
   @impl true
-  def connect(_params, socket, _connect_info) do
-    {:ok, socket}
+  def connect(%{"token" => encrypted_secret} = attrs, socket, connect_info) do
+    %{user_agent: user_agent, peer_data: %{address: remote_ip}} = connect_info
+
+    config = fetch_config!()
+    key_base = Keyword.fetch!(config, :key_base)
+    salt = Keyword.fetch!(config, :salt)
+    max_age = Keyword.fetch!(config, :max_age)
+
+    attrs =
+      attrs
+      |> Map.take(~w[external_id name_suffix public_key])
+      |> Map.put("last_seen_user_agent", user_agent)
+      |> Map.put("last_seen_remote_ip", remote_ip)
+
+    with {:ok, {id, secret}} <-
+           Plug.Crypto.verify(key_base, salt, encrypted_secret, max_age: max_age),
+         {:ok, token} <- Gateways.use_token_by_id_and_secret(id, secret),
+         {:ok, gateway} <- Gateways.upsert_gateway(token, attrs) do
```

Comment:
(we will want to track connected IPs from clients, relays, gateways, and admin logins)

Reply:
A separate table, but I keep it just for future UI/quick debugging use cases. The same will go with "who created and last updated this record" for everything in our system. Querying the Audit Logs table will be expensive since it tends to grow really fast.